yarn/src/main/scala/org/apache/spark/deploy/yarn/security/CredentialUpdater.scala

```diff
@@ -53,50 +53,58 @@ private[spark] class CredentialUpdater(
       override def run(): Unit = Utils.logUncaughtExceptions(updateCredentialsIfRequired())
     }
 
-  @volatile private var timeOfNextUpdate = sparkConf.get(CREDENTIALS_UPDATE_TIME)
+  /** Start the credentail updater thread periodically to get the credentials */
+  def bootstrap(): Unit = {
+    val bootstrapTime = sparkConf.get(CREDENTIALS_RENEWAL_TIME)
+    val remainingTime = bootstrapTime - System.currentTimeMillis()
+    if (remainingTime <= 0) {
+      // We just checked for new credentials but none were there, wait a minute and retry.
+      // This handles the shutdown case where the staging directory may have been removed(see
+      // SPARK-12316 for more details).
+      credentialUpdater.schedule(credentialUpdaterRunnable, 1, TimeUnit.MINUTES)
+    } else {
+      logInfo(s"Scheduling credentials refresh from HDFS in $remainingTime millis.")
+      credentialUpdater.schedule(credentialUpdaterRunnable, remainingTime, TimeUnit.MILLISECONDS)
+    }
+  }
 
-  def updateCredentialsIfRequired(): Unit = {
-    try {
+  private def updateCredentialsIfRequired(): Unit = {
+    val timeToNextUpdate = try {
       val credentialsFilePath = new Path(credentialsFile)
       val remoteFs = FileSystem.get(freshHadoopConf)
       SparkHadoopUtil.get.listFilesSorted(
         remoteFs, credentialsFilePath.getParent,
         credentialsFilePath.getName, SparkHadoopUtil.SPARK_YARN_CREDS_TEMP_EXTENSION)
-        .lastOption.foreach { credentialsStatus =>
+        .lastOption.map { credentialsStatus =>
         val suffix = SparkHadoopUtil.get.getSuffixForCredentialsPath(credentialsStatus.getPath)
```

Comment:
Doesn't look like your fault, but this whole block code is indented incorrectly. Should be one level deeper.